### PR TITLE
Filter only when necessary

### LINKED ISSUE #32: Filter cascades create too complex SQL
With one filter, everything seems fine. When a second filter is added, the first filter is applied redundantly (via joins) for every table in the data model. This may lead to an overload of the query planner.

Potential solutions:

- Add a new verb for applying referential integrity constraints, don't apply them yet after each filter
- Materialize enumerations of IDs for each table
- ?

``` r
library(tidyverse)
library(dm)

dm <-
  cdm_nycflights13() %>%
  cdm_select_tbl(-weather) %>%
  cdm_filter(flights, month == 1, day == 1) %>%
  cdm_copy_to(dbplyr::src_memdb(), .)

dm %>%
  cdm_filter(flights, year == 1) %>%
  cdm_get_tables() %>%
  map(dbplyr::sql_render)
#> $airlines
#> <SQL> SELECT * FROM `airlines` AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `RHS`
#>   WHERE (`LHS`.`carrier` = `RHS`.`carrier`)
#> )
#> 
#> $airports
#> <SQL> SELECT * FROM `airports` AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `RHS`
#>   WHERE (`LHS`.`faa` = `RHS`.`origin`)
#> )
#> 
#> $flights
#> <SQL> SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)
#> 
#> $planes
#> <SQL> SELECT * FROM `planes` AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `RHS`
#>   WHERE (`LHS`.`tailnum` = `RHS`.`tailnum`)
#> )

dm %>%
  cdm_filter(flights, year == 1) %>%
  cdm_filter(airlines, name == "xyz") %>%
  cdm_get_tables() %>%
  map(dbplyr::sql_render)
#> $airlines
#> <SQL> SELECT *
#> FROM (SELECT * FROM `airlines` AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `RHS`
#>   WHERE (`LHS`.`carrier` = `RHS`.`carrier`)
#> ))
#> WHERE (`name` = 'xyz')
#> 
#> $airports
#> <SQL> SELECT * FROM (SELECT * FROM `airports` AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `RHS`
#>   WHERE (`LHS`.`faa` = `RHS`.`origin`)
#> )) AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT * FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM (SELECT * FROM `airlines` AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `RHS`
#>   WHERE (`LHS`.`carrier` = `RHS`.`carrier`)
#> ))
#> WHERE (`name` = 'xyz')) AS `RHS`
#>   WHERE (`LHS`.`carrier` = `RHS`.`carrier`)
#> )) AS `RHS`
#>   WHERE (`LHS`.`faa` = `RHS`.`origin`)
#> )
#> 
#> $flights
#> <SQL> SELECT * FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM (SELECT * FROM `airlines` AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `RHS`
#>   WHERE (`LHS`.`carrier` = `RHS`.`carrier`)
#> ))
#> WHERE (`name` = 'xyz')) AS `RHS`
#>   WHERE (`LHS`.`carrier` = `RHS`.`carrier`)
#> )
#> 
#> $planes
#> <SQL> SELECT * FROM (SELECT * FROM `planes` AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `RHS`
#>   WHERE (`LHS`.`tailnum` = `RHS`.`tailnum`)
#> )) AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT * FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM (SELECT * FROM `airlines` AS `LHS`
#> WHERE EXISTS (
#>   SELECT 1 FROM (SELECT *
#> FROM `flights`
#> WHERE (`year` = 1.0)) AS `RHS`
#>   WHERE (`LHS`.`carrier` = `RHS`.`carrier`)
#> ))
#> WHERE (`name` = 'xyz')) AS `RHS`
#>   WHERE (`LHS`.`carrier` = `RHS`.`carrier`)
#> )) AS `RHS`
#>   WHERE (`LHS`.`tailnum` = `RHS`.`tailnum`)
#> )
```

<sup>Created on 2019-08-14 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0)</sup>

### PR DESCRIPTION
**Pull request by TSchiefer**
- store filter conditions in `dm`
- execute the necessary ones when retrieving a table (`tbl(dm, "table_name")`)
- execute all of them and update `dm` with `cdm_apply_filter_cascades()`

closes #32

### CODE REVIEW

File: R/dm.R [hunk 1]
```diff
@@ -370,7 +392,43 @@ tbl.dm <- function(src, from, ...) {
...
   check_correct_input(dm, from)
 
-  cdm_get_tables(dm)[[from]]
+  cdm_get_filtered_table(dm, from)
+}
+
+#' Apply all filters
+#'
+#' All set filters are applied and their combined cascading effect on each table of the [`dm`] is taken into account.
+#' An alias for this function is the method for `dm` class objects for `compute()`.
```
[the remark this answers is not in the document]
Comment (TSchiefer) [reply R/dm.R]: changed it to: "This function is called by the `compute()` method for `dm` class objects."

File: R/filter-dm.R [hunk 2]
```diff
@@ -65,3 +74,99 @@ cdm_semi_join <- function(dm, table, reduced_table) {
   join_list <- calculate_join_list(dm, table_name)
   perform_joins(filtered_dm, join_list)
 }
+
+cdm_get_filtered_table <- function(dm, from) {
+
+  filters <- cdm_get_filter(dm)
+  if (nrow(filters) == 0) return(cdm_get_tables(dm)[[from]])
```
[the remark this answers is not in the document]
Comment (TSchiefer) [reply R/filter-dm.R]: Could you please explain the benefit of getting rid of it?

File: R/graph.R [hunk 3]
```diff
@@ -49,15 +49,15 @@ cdm_get_referencing_tables <- function(dm, table) {
...
 
-  if (!table_name %in% names(igraph::V(g))) {
+  if (!table_name %in% names(V(g))) {
     return(tibble(
       lhs = character(), rhs = character(), rank = integer(), has_father = logical()
     ))
   }
 
-  bfs <- igraph::bfs(g, table_name, father = TRUE, rank = TRUE, unreachable = FALSE)
+  bfs <- bfs(g, table_name, father = TRUE, rank = TRUE, unreachable = FALSE)
```
[the remark this answers is not in the document]
Comment (TSchiefer) [reply R/graph.R]: alright, importing now only `E()` and `V()` explicitly

File: R/filter-dm.R (hunk 2 above)
[the remark this answers is not in the document]
Comment (krlmlr) [reply R/filter-dm.R]: One line of code less. What's more, reduced code complexity -- fewer code branches, fewer things to think about.

This does require that the code below works for the corner case of an empty set or zero-row table in various places -- in general we should strive towards this.

**Pull request merged**